src/filesystem.cpp

```diff
@@ -416,3 +416,44 @@ bool copy_file( const std::string &source_path, const std::string &dest_path )
 
     return dest_stream && source_stream;
 }
+
+std::string ensure_valid_file_name( const std::string &file_name, bool *change_made,
+                                    const std::string &invalid_chars, const char replacement_char )
+{
+    if( change_made != nullptr ) {
+        *change_made = false;
+    }
+    if( file_name.empty() ) {
+        return file_name;
+    }
+
+    /*
+     *  check if there's something wrong with the inputs.
+     */
+    //check if no invalid chars.
+    std::string tmp_invalid_chars = invalid_chars;
+    if( invalid_chars.empty() ) {
+        // default to common invalid characters for most file systems.
+        tmp_invalid_chars = "\\/:?\"<>|";
+    }
+
+    // check if replacement char is in the invalid chars...
+    char tmp_replacement = replacement_char;
+    if( invalid_chars.find( replacement_char ) != std::string::npos ) {
+        // default to space
+        tmp_replacement = ' ';
+    }
+
+    // do any replacement in the file name, if needed.
+    std::string new_file_name = file_name;
+    for( std::string::iterator it = new_file_name.begin(); it < new_file_name.end(); ++it ) {
```

Comment:
This could be a range-based for loop.

Reply:
Reworked the code in a more c++-idiomatic way.